sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+}
+
+static int bcmemmc_write(device_t *cdev, const void *buf, size_t len,
+                         size_t *wrote) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  device_t *emmcdev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)emmcdev->state;
+  resource_t *emmc = state->emmc;
+
+  assert(is_aligned(len, 4)); /* Assert multiple of 32 bits */
+
+  /* A very simple transfer */
+  for (size_t i = 0; i < len / sizeof(uint32_t); i++)
+    b_out(emmc, BCMEMMC_DATA, ((uint32_t *)buf)[i]);
+
+  /* TODO (mohr): check wether the transfer fully succeeded! */
+  if (wrote)
+    *wrote = len;
+  return 0;
+}
+
+#define GPHEN1 0x0068
+
+static void emmc_gpio_init(device_t *dev) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *gpio = state->gpio;
+  int64_t r = 0;
+  /* GPIO_CD */
+  bcm2835_gpio_function_select(gpio, 47, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_set_pull(gpio, 47, 2);
+  r = b_in(gpio, GPHEN1);
+  r |= 1 << 15;
+  b_out(gpio, GPHEN1, r);
+
+  /* GPIO_CLK, GPIO_CMD */
+  bcm2835_gpio_function_select(gpio, 48, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 49, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_set_pull(gpio, 48, 2);
+  bcm2835_gpio_set_pull(gpio, 49, 2);
+
+  /* GPIO_DAT0, GPIO_DAT1, GPIO_DAT2, GPIO_DAT3 */
+  bcm2835_gpio_function_select(gpio, 50, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 51, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 52, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 53, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_set_pull(gpio, 50, 2);
+  bcm2835_gpio_set_pull(gpio, 51, 2);
+  bcm2835_gpio_set_pull(gpio, 52, 2);
+  bcm2835_gpio_set_pull(gpio, 53, 2);
+}
+
+#define BCMEMMC_INIT_FREQ 400000
+
+static int bcmemmc_init(device_t *dev) {
+  assert(dev->driver == (driver_t *)&bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  int64_t r, cnt;
+
+  state->host_version =
+    (b_in(emmc, BCMEMMC_SLOTISR_VER) & HOST_SPEC_NUM) >> HOST_SPEC_NUM_SHIFT;
+  klog("e.MMC: GPIO set up");
+  /* Reset the card. */
+  b_out(emmc, BCMEMMC_CONTROL0, 0);
+  b_set(emmc, BCMEMMC_CONTROL1, C1_SRST_HC);
+  cnt = 10000;
+  do {
+    delay(30); /* ! */
+  } while ((b_in(emmc, BCMEMMC_CONTROL1) & C1_SRST_HC) && cnt--);
+  if (cnt <= 0) {
+    klog("ERROR: failed to reset EMMC");
+    return ETIMEDOUT;
+  }
+  klog("e.MMC: reset OK");
+  b_set(emmc, BCMEMMC_CONTROL1, C1_CLK_INTLEN | C1_TOUNIT_MAX);
+  /* Set clock to setup frequency. */
+  if ((r = bcmemmc_clk(dev, BCMEMMC_INIT_FREQ)))
+    return r;
+  b_out(emmc, BCMEMMC_INT_EN,
+        INT_CMD_DONE | INT_DATA_DONE | INT_READ_RDY | INT_WRITE_RDY |
+          INT_CMD_TIMEOUT | INT_DATA_TIMEOUT);
+  b_out(emmc, BCMEMMC_INT_MASK,
+        INT_CMD_DONE | INT_DATA_DONE | INT_READ_RDY | INT_WRITE_RDY |
+          INT_CMD_TIMEOUT | INT_DATA_TIMEOUT);
+
+  return 0;
+}
+
+static int bcmemmc_probe(device_t *dev) {
+  return dev->unit == 3;
```

Comment:
Now this is extremely interesting. Can you please explain how the whole driver can ever be attached as there is no relevant code in rootdev to create a new child and assign relevant resources (`rootdev_attach`)?

Reply:
Of course there isn't. The driver isn't even compiled into the code here, because of that quirky behaviour of linker sets that makes it impossible to create an empty device class. This means that until there's no SD Card driver in the source, which so far is the only driver present in the `emmc` device class, this driver cannot be compiled.

Reply:
Ok.